### PR TITLE
Check clang format

### PR DESCRIPTION
**Pull request by mgeplf**
* github action checks on push that changes are formatted correctly
* applied clang format to full all code

### CODE REVIEW

**Comment by mgeplf** [comment]
I haven't changed any C++ code, and the build is failing...

File: binds/python/bind_immutable.cpp [hunk 1]
```diff
@@ -1,7 +1,8 @@
+#include <pybind11/pybind11.h>
```
Comment (tristan0x) [inline binds/python/bind_immutable.cpp]: I am sure a good Samaritan will break the build by revert this.
Let me try to disable the formatting of the line before merging this.
https://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code

Comment (tristan0x) [reply binds/python/bind_immutable.cpp]: It is only possible to disable formatting within a range. Nothing elegant for a single line.

```diff
diff --git a/binds/python/bind_immutable.cpp b/binds/python/bind_immutable.cpp
index b9209a3..f12909d 100644
--- a/binds/python/bind_immutable.cpp
+++ b/binds/python/bind_immutable.cpp
@@ -1,5 +1,6 @@
+// clang-format off
 #include <pybind11/pybind11.h>
-
+// clang-format on
 #include <pybind11/iostream.h>
 #include <pybind11/numpy.h>
 #include <pybind11/operators.h>
```

Would this work for you @mgeplf ?

If not, IMO we should at least document the rationale:
```diff
diff --git a/binds/python/bind_immutable.cpp b/binds/python/bind_immutable.cpp
index b9209a3..9366ca3 100644
--- a/binds/python/bind_immutable.cpp
+++ b/binds/python/bind_immutable.cpp
@@ -1,3 +1,4 @@
+// top-level pybind11 header must be the included beforehand
 #include <pybind11/pybind11.h>
 
 #include <pybind11/iostream.h>
```

Comment (mgeplf) [reply binds/python/bind_immutable.cpp]: yeah, good call; I think I also need to clean up how this works b/c of the includes of `.cpp` and such files, but I'll do that in another commit

Comment (mgeplf) [reply binds/python/bind_immutable.cpp]: I should note that I *think* that clang format only sorts within blocks; so if there is a line break, it shouldn't be reverted.  I'll leave a comment though

Comment (tristan0x) [reply binds/python/bind_immutable.cpp]: > I should note that I _think_ that clang format only sorts within blocks; so if there is a line break, it shouldn't be reverted.

Absolutely.

> I'll leave a comment though

:+1:

File: ci/check_clang_format.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    "$VENV/bin/pip" install clang-format=="$CLANG_FORMAT_VERSION"
+fi
+
+set +u  # ignore errors in virtualenv's activate
+source "$VENV/bin/activate"
+set -u
+
+changes=$(git-clang-format 'HEAD~1')
+if [[ $(echo "$changes" | grep -n1 'changed files') ]]; then
+    echo "The following files require changes to pass the current clang-format"
```
Comment (wizmer) [inline ci/check_clang_format.sh]: Can you specify what action the user should take when this happen.
I guess it is to run the thing locally and commit ?

Comment (mgeplf) [reply ci/check_clang_format.sh]: pretty much

**Pull request merged**